RZDataBinding/RZDBCoalesce.m

```diff
@@ -0,0 +1,309 @@
+            [notification send];
+        }];
+    }
+}
+
++ (void)coalesceBlock:(void (^)())coalesceBlock
+{
+    NSParameterAssert(coalesceBlock);
+
+    [RZDBCoalesce begin];
+    coalesceBlock();
+    [RZDBCoalesce commit];
+}
+
++ (void)setAutoCoalesceByRunloopOnMainThread:(BOOL)autoCoalesce
+{
+    objc_setAssociatedObject(self, @selector(autoCoalesceByRunloopOnMainThread), @(autoCoalesce), OBJC_ASSOCIATION_RETAIN_NONATOMIC);
+}
+
+#pragma mark - private methods
+
++ (BOOL)autoCoalesceByRunloopOnMainThread
+{
+    return [objc_getAssociatedObject(self, _cmd) boolValue];
+}
+
++ (RZDBCoalesce *)currentCoalesce
+{
+    return [NSThread currentThread].threadDictionary[kRZDBCoalesceStorageKey];
+}
+
++ (RZDBCoalesce *)currentCoalesceBeginIfNeeded:(BOOL)begin
+{
+    RZDBCoalesce *current = [self currentCoalesce];
+
+    if ( current == nil && [NSThread isMainThread] && [self autoCoalesceByRunloopOnMainThread] ) {
+        [RZDBCoalesce begin];
+        [self performSelector:@selector(commit) withObject:nil afterDelay:0.0 inModes:@[NSRunLoopCommonModes]];
+
+        current = [self currentCoalesce];
+    }
+
+    return current;
+}
+
++ (void)setCurrentCoalesce:(RZDBCoalesce *)current
+{
+    if ( current != nil ) {
+        [NSThread currentThread].threadDictionary[kRZDBCoalesceStorageKey] = current;
+    }
+    else {
+        [[NSThread currentThread].threadDictionary removeObjectForKey:kRZDBCoalesceStorageKey];
+    }
+}
+
+- (instancetype)init
+{
+    self = [super init];
+    if ( self ) {
+        _beginCount = 1;
+        _notifications = [NSMutableArray array];
+    }
+
+    return self;
+}
+
+- (void)addNotification:(RZDBNotification *)notification
+{
+    NSUInteger existingIdx = [self.notifications indexOfObjectPassingTest:^BOOL(RZDBNotification *existing, NSUInteger idx, BOOL *stop) {
+        return [notification isEqual:existing];
+    }];
+
+    if ( existingIdx != NSNotFound ) {
+        RZDBNotification *existing = self.notifications[existingIdx];
+
+        if ( notification.changeDict[kRZDBChangeKeyNew] != nil ) {
+            existing.changeDict[kRZDBChangeKeyNew] = notification.changeDict[kRZDBChangeKeyNew];
+        }
+        else {
+            [existing.changeDict removeObjectForKey:kRZDBChangeKeyNew];
+        }
+    }
+    else {
+        [self.notifications addObject:notification];
+    }
+}
+
+@end
+
+#pragma mark - RZDBNotification implementation
+
+@implementation RZDBNotification
+
++ (instancetype)notificationWithInvocation:(NSInvocation *)invocation
+{
+    RZDBNotification *notification = nil;
+
+    NSMethodSignature *methodSig = invocation.methodSignature;
+
+    // rzdb callbacks must have void return type, and
+    // if there are any arguments, the first and only argument must be a dictionary
+    if ( methodSig.methodReturnLength == 0 && (methodSig.numberOfArguments == 2 || methodSig.numberOfArguments == 3) ) {
```

Comment:
Discussed offline. `RZDBNotification` holds a `weak` reference to the target, whereas `NSInvocation` has an `__unsafe_unretained` reference. `weak` is preferable, in case the target gets deallocated before the coalesce has been committed.